ambry-account/src/main/java/com/github/ambry/account/HelixAccountService.java

```diff
@@ -0,0 +1,525 @@
+  static final String FULL_ACCOUNT_METADATA_CHANGE_MESSAGE = "full_account_metadata_change";
+  static final String ACCOUNT_METADATA_MAP_KEY = "accountMetadata";
+  static final String FULL_ACCOUNT_METADATA_PATH = "/account_metadata/full_data";
+  private static final String ZN_RECORD_ID = "full_account_metadata";
+  private final Logger logger = LoggerFactory.getLogger(getClass());
+  private final HelixPropertyStore<ZNRecord> helixStore;
+  private final AccountServiceMetrics accountServiceMetrics;
+  private final TopicListener<String> listener;
+  private final Notifier<String> notifier;
+  private final AtomicReference<AccountInfoMap> accountInfoMapRef = new AtomicReference<>(new AccountInfoMap());
+  private volatile boolean isOpen = false;
+
+  /**
+   * Constructor. It fetches the full account metadata set and caches locally.
+   * This call is blocking until it fetches all the {@link Account} metadata from {@link HelixPropertyStore}.
+   * @param helixStore A {@link HelixPropertyStore} used by the {@code HelixAccountService}. Cannot be {@code null}.
+   * @param accountServiceMetrics {@link AccountServiceMetrics} to report metrics. Cannot be {@code null}.
+   * @param notifier A {@link Notifier} that will be used to publish message after updating {@link Account}s, and
+   *                 listen to {@link Account} change messages. Cannot be {@code null}.
+   */
+  HelixAccountService(HelixPropertyStore<ZNRecord> helixStore, AccountServiceMetrics accountServiceMetrics,
+      Notifier<String> notifier) {
+    if (helixStore == null || accountServiceMetrics == null || notifier == null) {
+      throw new IllegalArgumentException("helixStore or accountServiceMetrics or notifier cannot be null");
+    }
+    this.helixStore = helixStore;
+    this.accountServiceMetrics = accountServiceMetrics;
+    this.notifier = notifier;
+    listener = (topic, message) -> {
+      checkOpen();
+      logger.trace("Start to process message={} for topic={}", message, topic);
+      boolean noException = true;
+      try {
+        switch (message) {
+          case FULL_ACCOUNT_METADATA_CHANGE_MESSAGE:
+            readFullAccountAndUpdateCache(FULL_ACCOUNT_METADATA_PATH, false);
+            break;
+
+          default:
+            logger.error("Could not understand message={} for topic={}", message, topic);
+            accountServiceMetrics.unRecognizedMessageErrorCount.inc();
+        }
+      } catch (Exception e) {
+        noException = false;
+        logger.error("Exception occurred when processing message={} for topic={}.", message, topic, e);
+        accountServiceMetrics.processFullAccountMessageErrorCount.inc();
+      }
+      if (noException) {
+        logger.trace("Completed processing message={} for topic={}", message, topic);
+      }
+    };
+    notifier.subscribe(ACCOUNT_METADATA_CHANGE_TOPIC, listener);
+    readFullAccountAndUpdateCache(FULL_ACCOUNT_METADATA_PATH, true);
```

Comment:
Discussed offline, the order is correct so I will keep it. Added a lock to protect the update method to be called sequentially.